File: src/main/java/com/codeborne/selenide/impl/commands/SetValue.java
package com.codeborne.selenide.impl.commands;

import com.codeborne.selenide.Command;
import com.codeborne.selenide.SelenideElement;
import com.codeborne.selenide.ex.InvalidStateException;
import com.codeborne.selenide.impl.WebElementSource;
import org.openqa.selenium.WebElement;

import static com.codeborne.selenide.Configuration.fastSetValue;
import static com.codeborne.selenide.Selenide.executeJavaScript;
import static com.codeborne.selenide.impl.Events.events;

public class SetValue implements Command<WebElement> {
  SelectOptionByValue selectOptionByValue = new SelectOptionByValue();
  SelectRadio selectRadio = new SelectRadio();
  
  @Override
  public WebElement execute(SelenideElement proxy, WebElementSource locator, Object[] args) {
    String text = (String) args[0];
    WebElement element = locator.findAndAssertElementIsVisible();
    if ("select".equalsIgnoreCase(element.getTagName())) {
      selectOptionByValue.execute(proxy, locator, args);
    }
    else if ("input".equalsIgnoreCase(element.getTagName()) && "radio".equals(element.getAttribute("type"))) {
      selectRadio.execute(proxy, locator, args);
    }
    else if (text == null || text.isEmpty()) {
      element.clear();
    }
    else if (fastSetValue) {
      text = truncateMaxLength(element, text);
      String error = executeJavaScript(
          "if (arguments[0].getAttribute('readonly') != undefined) " +
              "  return 'Cannot change value of readonly element';" +
              "arguments[0].value = arguments[1];" +
              "return null;", element, text);
      if (error != null) throw new InvalidStateException(error);
      events.fireEvent(element, "focus", "keydown", "keypress", "input", "keyup", "change");
    }
    else {
      element.clear();
      element.sendKeys(text);
      events.fireChangeEvent(element);
    }

    return proxy;
  }

  private String truncateMaxLength(WebElement element, String text) {
    try {
      String maxlength = element.getAttribute("maxlength");
      int elementMaxLength = Integer.parseInt(maxlength);
      return text.length() > elementMaxLength ? text.substring(0, elementMaxLength) : text;
    }
    catch (NumberFormatException invalidMaxLength) {
      return text;
    }
  }
}

File: src/main/resources/set-value.js
(function(webelement, text) {

  function triggerEvent(target, eventName) {
    if (document.createEventObject) {
      var event = document.createEventObject();
      target.fireEvent('on' + eventName, event);
    }
    else {
      var event = document.createEvent('HTMLEvents');
      event.initEvent(eventName, true, true);
      target.dispatchEvent(event);
    }
  }

  function trigger(target, eventNames) {
    for (var i in eventNames) {
      try {
        triggerEvent(target, eventNames[i]);
      }
      catch (staleElementException) {
        console.log('failed to trigger event', eventNames[i])
      }
    }
  }

  if (webelement.getAttribute('readonly') !== null) return 'Cannot change value of readonly element';
  if (webelement.getAttribute('disabled') !== null) return 'Cannot change value of disabled element';

  trigger(document.activeElement, ['blur']);
  webelement.focus();
  var maxlength = webelement.getAttribute('maxlength') == null ? -1 : parseInt(webelement.getAttribute('maxlength'));
  webelement.value = maxlength === -1 ? text : text.length <= maxlength ? text : text.substring(0, maxlength);
  trigger(webelement, ['focus', 'keydown', 'keypress', 'input', 'keyup', 'change']);

  return "";
})(arguments[0], arguments[1]);

